### PR TITLE
provisioner: replace curl calls to crowbar API with crowbarctl

### PR DESCRIPTION
**Pull request by jsuchome**
This replacement will fix some issues in relation with curl,
https, redirections and BasicAuth passwords."

**Why is this change necessary?**

This is part of ongoing work to use crowbarctl instead of direct curl calls to API. crowbarctl should know everything and no special options that we're building for curl command are necessary..

**Trello card**

https://trello.com/c/2SwQTBg1/399-p3-replace-curl-commands-in-crowbar-with-crowbarctl

**Requires**

sleshammer changes:
https://build.suse.de/package/show/home:jsuchome:branches:Devel:Cloud:8/sleshammer# -> https://build.suse.de/request/show/154768

### CODE REVIEW

**Comment by jsuchome** [comment]
[P3 Replace curl commands in crowbar with crowbarctl](https://trello.com/c/2SwQTBg1/399-p3-replace-curl-commands-in-crowbar-with-crowbarctl)

**Comment by susecloudcommitpolice** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please limit commit subject line to 50 characters.<br />
8e675345b137b76ad27dc39fbfa58a1f6fe3442d</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: bin/gather_logs.sh [hunk 1]
```diff
@@ -47,11 +47,9 @@ sort_by_last() {
...
     logs=(/var/log /etc)
     logs+=(/var/chef/cache /var/cache/chef /opt/dell/crowbar_framework/db)
-    curlargs=(-o /dev/null -D - --connect-timeout 30 --max-time 120 --insecure \
-    --location)
-    [[ $CROWBAR_KEY ]] && curlargs+=(--digest -u "$CROWBAR_KEY")
-    for to_get in nodes proposals roles; do
-	curl "${curlargs[@]}" "http://localhost/$to_get" || :
+    crowbarctl node list
+    for to_get in proposals roles; do
+        crowbarctl $to_get proposal list crowbar
```
Comment (jsuchome) [inline bin/gather_logs.sh]: This is of course not 1:1 change, but I do not really understand the original meaning of those calls ...is it to just generate some activity in the log files?

File: chef/cookbooks/provisioner/templates/default/autoyast.xml.erb [hunk 2]
```diff
@@ -304,7 +303,7 @@ curl -s -o /mnt/usr/sbin/crowbar_join <%= @crowbar_join %>
 chmod +x /mnt/usr/sbin/crowbar_join
 
 mkdir -p /mnt/var/log/crowbar
-post_state $HOSTNAME "installed"
+crowbarctl node transition $HOSTNAME "installed"
```
Comment (jsuchome) [inline chef/cookbooks/provisioner/templates/default/autoyast.xml.erb]: OK, this does not work this way. It's not possible to simple call `/mnt/usr/bin/crowbarctl` due to missing dependencies.

It might be possible if the post-script itself is chrooted  (`<chrooted config:type="boolean">false</chrooted>`) and I do not know why don't we have it...

Comment (jsuchome) [reply chef/cookbooks/provisioner/templates/default/autoyast.xml.erb]: Or I might add extra chrooted script that will only call crowbarctl from the new installed system...

Comment (nicolasbock) [reply chef/cookbooks/provisioner/templates/default/autoyast.xml.erb]: Have you tested this and is it working? Your comment seems to suggest that it's not quite there yet :smile:

Comment (jsuchome) [reply chef/cookbooks/provisioner/templates/default/autoyast.xml.erb]: Oh - the comment is obsolete, I've addressed this in 2nd commit.

Comment (nicolasbock) [reply chef/cookbooks/provisioner/templates/default/autoyast.xml.erb]: Oh, sorry. That's sometimes a little hard to tell on GitHub.

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb [hunk 3]
```diff
@@ -64,25 +64,18 @@
...
-                -L -X POST --data-binary "{ \"name\": \"$1\", \"state\": \"$2\" }" \
-                -H "Accept: application/json" -H "Content-Type: application/json" \
-                --max-time 240)
-              [[ $CROWBAR_KEY ]] && curlargs+=(-u "$CROWBAR_KEY" --digest --anyauth)
-              curl "${curlargs[@]}" "http://$IP/crowbar/crowbar/1.0/transition/default"
-          }
 
-          curl -s -o /mnt/usr/sbin/crowbar_join <%= @crowbar_join %>
-          chmod +x /mnt/usr/sbin/crowbar_join
+          export PASS="$(sed -e 's/^machine-install://' <<< $CROWBAR_KEY)"
```
Comment (vuntz) [inline chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Do we need this line?

File: chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb [hunk 4]
```diff
@@ -506,8 +490,8 @@ if [ "$MODE" == "setup" -o "$MODE" == "start" ]; then
 
     [ $final_state == "ready" ] || EXVAL=1
 elif [ "$MODE" == "stop" ]; then
-    HOSTNAME=$(hostname -f)
     state=$(get_state $HOSTNAME)
+    HOSTNAME=$(hostname -f)
```
Comment (vuntz) [inline chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb]: Why this change?

File: chef/cookbooks/provisioner/templates/suse/crowbar_register.erb [hunk 5]
```diff
@@ -246,8 +206,7 @@ fi
 # Check that we can really register this node
 # -------------------------------------------
 
-nodes=$(get_nodes_allow_fail || echo "")
-if test -z "$nodes"; then
+if ! crowbarctl node list > /dev/null; then
```
Comment (vuntz) [inline chef/cookbooks/provisioner/templates/suse/crowbar_register.erb]: That one is wrong, and tricky. When people use crowbar_join, you can't expect crowbarctl to be installed on the node. It's actually a package that should be added to the list of things to install a few lines later (look for zypper in this script).

One solution would be to first install crowbarctl, and then do the checks like this.

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 3 above)
Comment (jsuchome) [reply chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Oh, I forgot the block with writing `/etc/crowbarrc`. Or could we already assume it is there when `autoyast-upgrade.xml.erb` is used?

File: chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb (hunk 4 above)
Comment (jsuchome) [reply chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb]: Because `crowbarcrl node` (that gets executed in `get_state`) does not expect node's fqdn AFAIK

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 3 above)
Comment (vuntz) [reply chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: So I was wondering about the upgrade case. Not sure how it's handled. Maybe this is the right place?

File: chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb (hunk 4 above)
Comment (vuntz) [reply chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb]: Ah, it may make sense to add a comment about it for the future.

File: chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb (hunk 3 above)
Comment (jsuchome) [reply chef/cookbooks/provisioner/templates/default/autoyast-upgrade.xml.erb]: Yeah, so I've added `/etc/crowbarrc` here as well, just to be sure it is present.

**Comment by jsuchome** [comment]
@rsalevsky @vuntz any chance for another review round?

File: chef/cookbooks/provisioner/templates/default/net-post-install.sh.erb [hunk 6]
```diff
@@ -93,7 +84,7 @@ apt-get -y --force-yes remove apparmor
 apt-get -y --force-yes install screen
 sleep 30
 
-post_state $HOSTNAME "installed"
+crowbarctl node transition $HOSTNAME "installed"
```
Comment (JanZerebecki) [inline chef/cookbooks/provisioner/templates/default/net-post-install.sh.erb]: According to my reading /etc/crowbarrc is not set up yet, how will this work?

File: chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb [hunk 7]
```diff
@@ -65,35 +65,19 @@ log_to() {
...
   local output
-  local curlargs=(--connect-timeout 60 -s -L \
-      -H "Accept: application/json" -H "Content-Type: application/json" \
-      --max-time 240 --insecure --location)
-  [[ $CROWBAR_KEY ]] && curlargs+=(-u "$CROWBAR_KEY" --digest --anyauth)
-  # we expect something like {"value":"ready"}
-  output=$(curl "${curlargs[@]}" "http://$IP/dashboard/$1/attribute/state" | \
-      tr -d "{}\"'[:space:][:cntrl:]")
+  # we expect something like "status": "Ready"
+  output=$(crowbarctl node status --json --filter $1 | grep status | \
```
Comment (JanZerebecki) [inline chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb]: I'm unsure about if /etc/crowbarrc is ensured to be present at this point, do you know?

File: chef/cookbooks/provisioner/templates/default/compute.ks.erb [hunk 8]
```diff
@@ -140,7 +131,7 @@ sed -i -e '/initdefault/ s/5/3/' /etc/inittab
 
 sleep 30
 
-post_state $HOSTNAME "installed"
+crowbarctl node transition $HOSTNAME "installed"
```
[the remark this answers is not in the document]
Comment (jsuchome) [reply chef/cookbooks/provisioner/templates/default/compute.ks.erb]: Hm, now I see `compute.ks.erb` is for redhat/centos nodes... I assume we do not install crowbarctl there so I should keep curl calls I think

File: chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb (hunk 7 above)
Comment (jsuchome) [reply chef/cookbooks/provisioner/templates/default/crowbar_join.suse.sh.erb]: I believe at the time of crowbar_join calls `/etc/crowbarrc` is already installed (either the temporary one created by autoyast, or the real one created by `recipes/base.rb`
